### PR TITLE
Validations model specs

### PR DESCRIPTION
**Pull request by Kaiserabbas**
# Added the the validations:
- For the User model:
    - Name must not be blank.
    - PostsCounter must be an integer greater than or equal to zero.
- For the Post model:
    - Title must not be blank.
    - Title must not exceed 250 characters.
    - CommentsCounter must be an integer greater than or equal to zero.
    - LikesCounter must be an integer greater than or equal to zero.
- Added unit specs for all of your models' methods and validations.

### CODE REVIEW

File: spec/models/user_spec.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    user.valid?
+    expect(user.errors[:name]).to include("can't be blank")
+  end
+
+  it 'is not valid with a negative posts_counter' do
+    user = User.new(name: 'Tom', post_counter: -1)
+    user.valid?
+    expect(user.errors[:post_counter]).to include('must be greater than or equal to 0')
+  end
+
```
Comment (topeogunleye) [inline spec/models/user_spec.rb:20]: ---

Hello,

Great work on writing comprehensive tests for the User model! Here are some insights based on the project requirements and the code you've provided:

✅ **Positive Testing:**
- You've correctly tested that a user with a name and a non-negative `posts_counter` is valid. This aligns well with the validation requirements.

✅ **Negative Testing:**
- You've included tests to verify that a user without a name is not valid, as well as a test to ensure that the `posts_counter` cannot be negative. These are crucial negative test cases that properly validate model constraints.

🛠️ **Recommendations:**
- Include a test case to check if `posts_counter` is an integer. You can write a test that assigns a float or string to `posts_counter` and expects it not to be valid.

- Make sure that you have the actual validations in your User model file (`user.rb`). The tests will pass only if the validations are present in the model.

```ruby
# app/models/user.rb
class User < ApplicationRecord
  validates :name, presence: true
  validates :posts_counter, numericality: { only_integer: true, greater_than_or_equal_to: 0 }
end
```

- Don't forget to test other scenarios if required by the project specifications, like validations for other attributes, associations, or class and instance methods.

Keep up the good work! It's evident that you're applying TDD principles effectively, which is a key skill in software development.

---

File: spec/models/post_spec.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  it 'updates the posts counter for a user' do
+    user = User.create(name: 'Tom', post_counter: 0)
+    post = user.posts.create(title: 'My Post', comments_counter: 0, likes_counter: 0)
+
+    post.increment_user_post_counter
+
+    user.reload
+    expect(user.post_counter).to eq(2)
+  end
+
```
Comment (topeogunleye) [inline spec/models/post_spec.rb:75]: ---

Hello,

You've done an excellent job testing the Post model and ensuring that it adheres to the validation requirements outlined for the project. Here's a breakdown of the review based on your code:

🛠️ **Recommendations:**
- It's not clear if you are testing the datatype of `comments_counter` and `likes_counter` to ensure they are integers. Consider adding tests to confirm that these fields reject non-integer values.

- In your model (`post.rb`), ensure you have the following validations to pass the tests:

```ruby
# app/models/post.rb
class Post < ApplicationRecord
  validates :title, presence: true, length: { maximum: 250 }
  validates :comments_counter, numericality: { only_integer: true, greater_than_or_equal_to: 0 }
  validates :likes_counter, numericality: { only_integer: true, greater_than_or_equal_to: 0 }
end
```

Overall, you're demonstrating a solid understanding of RSpec and Rails model validations. Keep up the great work, and remember to align your tests with the exact behavior of your methods!

---

**Review by topeogunleye — CHANGES_REQUESTED** [review]
Hi @Kaiserabbas, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

Try-Again

> ## (Highlights) Good Points: 💪 
✅ **Positive Testing:**
- You've correctly tested that a user with a name and a non-negative `posts_counter` is valid. This aligns well with the validation requirements.

✅ **Negative Testing:**
- You've included tests to verify that a user without a name is not valid, as well as a test to ensure that the `posts_counter` cannot be negative. These are crucial negative test cases that properly validate model constraints.

✅ **Positive Validation Tests:**
- Your test for the post being valid with the appropriate title, comments_counter, and likes_counter is on point. It ensures the model is valid when all conditions are met.

✅ **Presence Validation Test:**
- The test for the title presence with `it 'is not valid without a title'` correctly captures the scenario where a post must have a title.

✅ **Length Validation Test:**
- You've successfully tested the maximum length of the title with `it 'is not valid with a title exceeding 250 characters'`, aligning perfectly with the requirements.

✅ **Numericality Validation Tests:**
- The tests for `comments_counter` and `likes_counter` being non-negative are well-written and validate that both counters cannot be less than zero.

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- For future pull requests, please notice that your branch name `validations-model-specs` does not follow the convention of being named after a feature or action. For example, you could name your branch `added-validations-model-specs`. This makes it easier to track the changes and the purpose of each branch.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag @topeogunleye in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Kaiserabbas** [comment]
Thank you for the review and suggesstions. Add new tests as per reviewer's comments

**Review by AmaduKamara — APPROVED** [review]
# Status: APPROVED  🟢 🟢 🟢

Hi @Kaiserabbas,

Good job working on the changes highlighted by the previous reviewer 👍🏽 

Your project is complete!  There is nothing else to say other than... it's time to merge it: ship it:
Congratulations! 🎉 

## Highlights:

✅ You set up Gitflow correctly 👍🏽
✅ The Linter checks passed successfully 👍🏽
✅ Good commit messages and PR description 👍🏽
✅ Set up validation and model specs correctly 👍🏽
✅ All checks passed successfully as expected 👍🏽
✅ Excellent communication with reviewer 👍🏽

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
```

**Pull request merged**